### ISSUE TITLE
opt_level='O1' wrong loading checkpoint behavior when loaded after evaluate function 

### ISSUE THREAD

**Issue by Synicix** [NVIDIA/apex#796]
There seems to be a issue with amp checkpoint loading issue when amp is set to opt_level='O1'.

It seems to occur if the code logic follow this: 
- Train -> Save_model(epoch=5) -> Train -> Evaluate() -> Load_model(epoch=5) -> Evaluate()

The output is: 
`Test Score at epoch 5 : 4.6053485631942745
Model Saved
Test Score at epoch 100 : 4.09333438873291
Model Loaded
Test Score at epoch 5 loaded from checkpoint: 4.09333438873291`

Which is wrong where it seem that the 

However, if I don't call evaluate before loading the model then incorrect model loading doesn't occur as in:
- Train -> Save_model(epoch=5) -> Train -> Load_model(epoch=5) -> Evaluate()

The output is:
`Test Score at epoch 5 : 4.6072866201400755
Model Saved
Model Loaded
Test Score at epoch 5 loaded from checkpoint: 4.6072866201400755`

Which is correct

Here is the reference code to reproduce the error:
`import torch.nn as nn
import torch.nn.functional as F

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 100)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

import torch
import torchvision
import torchvision.transforms as transforms
import numpy as np
from apex import amp
import time

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR100(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=256,
                                          shuffle=True, num_workers=4)

testset = torchvision.datasets.CIFAR100(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=256,
                                         shuffle=False, num_workers=4)

device = torch.device('cuda:0')

model = Net().to(device)
optimizer = torch.optim.Adam(model.parameters(), eps=1e-4)
criterion = nn.CrossEntropyLoss()

model, optimizer = amp.initialize(model, optimizer, opt_level='O1')

print(model)

# Evaluate function
def evaluate():
    model.eval()
    
    with torch.no_grad():
        running_loss = []
        for data in testloader:
            inputs = data[0].to(device)
            targets = data[-1].to(device)

            outputs = model(inputs)
            loss = criterion(outputs, data[-1].to(device))
            running_loss.append(loss.item())
        
    model.train()
    return np.array(running_loss).mean()
    
    
# Save model function
def save_model():
    checkpoint = dict(model=model.state_dict(), optimizer=optimizer.state_dict(), amp=amp.state_dict())
    torch.save(checkpoint, 'temp.pth')
    print('Model Saved')
    
def load_model():
    checkpoint = torch.load('temp.pth', map_location=device)
    model.load_state_dict(checkpoint['model'])
    optimizer.load_state_dict(checkpoint['optimizer'])
    amp.load_state_dict(checkpoint['amp'])
    print('Model Loaded')

for i, data in enumerate(trainloader):
    inputs = data[0].to(device)
    targets = data[-1].to(device)
    
    outputs = model(inputs)
    loss = criterion(outputs, data[-1].to(device))
    with amp.scale_loss(loss, optimizer) as scale_loss:
        scale_loss.backward()
    
    optimizer.step()
    optimizer.zero_grad()
    
    # break out after 10 epochs
    if (i + 1) == 5:
        print('Test Score at epoch ' + str(i + 1) + ' : ' + str(evaluate()))
        save_model()
        
    if (i + 1) == 50:
        # Line to comment out to see the other case
        print('Test Score at epoch ' + str(i + 1) + ' : ' + str(evaluate()))
        load_model()
        print('Test Score at epoch ' + str(5) + ' loaded from checkpoint: ' + str(evaluate()))
        break
`

**Comment by Synicix** [comment]
Note that for O0, O2, and O3, this doesn't occur
